File: src/algorithm/leetcode/Problem647.java
package algorithm.leetcode;

public class Problem647 {

    public static int countSubstrings(String s) {
        int[][] dp = new int[s.length()][s.length()];
        for (int i = 0; i < s.length(); i++)
            dp[i][i] = 1;
        for (int i = 0; i < s.length() - 1; i++) {
            if(s.charAt(i) == s.charAt(i+1))
                dp[i][i + 1] = 1;
        }
        for (int len = 3; len <= s.length(); len++) {
            for (int start = 0; start <= s.length() - len; start++) {
                if (s.charAt(start) == s.charAt(start + len - 1)) {
                    dp[start][start + len - 1] = dp[start + 1][start + len - 2];
                } else {
                    dp[start][start + len - 1] = 0;
                }
            }
        }
        int count = 0;
        for (int len = 1; len <= s.length(); len++) {
            for (int start = 0; start <= s.length() - len; start++) {
                if (dp[start][start + len - 1] == 1)
                    count++;
            }
        }
        return count;
    }

    public static void main(String[] args) {
        System.out.println(countSubstrings("aaa"));
    }

}

File: src/algorithm_cpp/problem647.cpp
class Solution {
public:
    int countSubstrings(string s) {
	int count = 0;
	vector<vector<int>> dp(s.size(), vector<int>(s.size(), 0));
	for (int i = 0; i < s.size(); i++) {
	    dp[i][i] = 1;
	    count++;
	}
	for (int i = 1; i < s.size(); i++) {
	    if(s.at(i - 1) == s.at(i)) {
	    dp[i - 1][i] = 1;
	    count++;
	}
	for (int len = 3; len <= s.size(); len++) {
	    for (int start = 0; start <= s.size() - len; start++) {
		if (s.at(start) == s.at(start + len - 1) && dp[start + 1][start + len - 2] == 1) {
		    dp[start][start + len - 1] = 1;
		    count++;
		}
	    }
	}
	return count;
    }
		
};
